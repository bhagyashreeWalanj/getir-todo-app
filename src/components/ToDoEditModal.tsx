import React, { useState } from 'react'
import Modal from 'react-modal'
import styles from '../styles/modal.module.scss'
import { RiCloseCircleFill } from 'react-icons/ri'
import Alert from './Alert'

interface IEditTaskModalProps {
  taskName: string
  isOpen: boolean
  onRequestClose: () => void
  handleEditTask: (id: string, title: string) => void
  EditTaskId: string
}

const ToDoEditModal = ({
  taskName,
  isOpen,
  onRequestClose,
  handleEditTask,
  EditTaskId,
}: IEditTaskModalProps) => {
  const [newTitle, setNewTitle] = useState('')
  const [alert, setAlert] = useState({ show: false, msg: '', type: '' })

  const showAlert = (show = false, type = '', msg = '') => {
    setAlert({ show, type, msg })
  }

  const handleEditOnChange = (title: string) => {
    setNewTitle(title)
  }
  const editTask = (id: string, title: string) => {
    if (title !== '') {
      handleEditTask(id, title)
      showAlert(true, 'success-edit', 'Successfully changed')
      const timeout = setTimeout(() => {
        onRequestClose()
        setAlert({ show: false, msg: '', type: '' })
      }, 2000)
      return () => clearTimeout(timeout)
    } else {
      showAlert(true, 'danger-edit', 'Please add task to save')
      return
    }
  }

  return (
    <div>
      <Modal
        isOpen={isOpen}
        onRequestClose={onRequestClose}
        className={styles.modalContainer}
        overlayClassName={styles.overlayContainer}
        ariaHideApp={false}
      >
        <RiCloseCircleFill
          className={styles.closeButton}
          onClick={() => {
            onRequestClose()
          }}
        />
        {alert.show && <Alert {...alert} removeAlert={showAlert} list={[]} />}

        {!!taskName && (
          <div key={EditTaskId}>
            <div className={styles.label}>
              <h2 className={styles.header}>Edit Task</h2>
            </div>

            <div className={styles.NoteTypeContainer}>
              <input
                type="text"
                className={styles.inputText}
                placeholder="Edit task..."
                //value={task}
                defaultValue={taskName}
                name={`input_${EditTaskId}`}
                onChange={(e) => handleEditOnChange(e.target.value)}
              />

              <button
                className={styles.ButtonSave}
                onClick={() => {
                  editTask(EditTaskId, newTitle)
                }}
              >
                <span>Save</span>
              </button>
            </div>
          </div>
        )}
      </Modal>
    </div>
  )
}

export default ToDoEditModal
