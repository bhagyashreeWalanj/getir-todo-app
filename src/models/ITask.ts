export interface ITask {
    id: string;
    title: string;
    createdAt: string;
    completed: boolean;
  }
  