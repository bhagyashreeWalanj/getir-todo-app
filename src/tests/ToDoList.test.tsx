import React from 'react'
import { shallow, mount } from 'enzyme'
import ToDoAppList from '../components/ToDoList'
import { ITask } from '../models/ITask'
import ToDoEditModal from '../components/ToDoEditModal'

describe('Testing ToDoList Component', () => {
  const editTask = jest.fn()
  const removeTask = jest.fn()
  const taskList: ITask[] = [
    {
      id: '1',
      title: 'Hello World 1',
      createdAt: '01/05/2022',
      completed: false,
    },
    {
      id: '2',
      title: 'Hello World 2',
      createdAt: '01/05/2022',
      completed: false,
    },
  ]

  const wrapper = shallow(
    <ToDoAppList
      editTask={editTask}
      removeTask={removeTask}
      tasks={taskList}
    />,
  )

  let pendingToDos = wrapper.find('.taskTR_pending')
  let completedToDos = wrapper.find('.taskTR_completed')

  it('should display to ToDo List Component', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have all input elements', () => {
    expect(wrapper.find('.ToDo__checkbox').length).toBeTruthy()
    expect(wrapper.find('.ToDo__update').length).toBeTruthy()
    expect(wrapper.find('.ToDo__delete').length).toBeTruthy()
  })

  it('should render All pending tasks', () => {
    expect(pendingToDos.exists()).toBeTruthy()
    expect(pendingToDos.length).toBe(2)
  })

  it('should render all completed tasks ', () => {
    expect(completedToDos.exists()).toBeFalsy()
    expect(completedToDos.length).toBe(0)
  })

  it('should render all tasks', () => {
    const totalTasks = completedToDos.length + pendingToDos.length
    expect(totalTasks).toBe(taskList.length)
  })

  it('should confirm checkbox is checked for pending task', () => {
    const wrapperMount = mount(
      <ToDoAppList
        {...taskList}
        tasks={taskList}
        editTask={editTask}
        removeTask={removeTask}
      />,
    )
    wrapperMount
      .find('#checkbox_1')
      .simulate('change', { target: { checked: true } })
    expect(wrapperMount.find('#checkbox_1').props().checked).toBe(true)
  })

  it('Should confirm checkbox is unchecked for completed task', () => {
    taskList[0].completed = true
    const wrapperMount = mount(
      <ToDoAppList
        {...taskList}
        tasks={taskList}
        editTask={editTask}
        removeTask={removeTask}
      />,
    )
    wrapperMount
      .find('#checkbox_1')
      .simulate('change', { target: { checked: false } })
    expect(wrapperMount.find('#checkbox_1').props().checked).toBe(false)
  })

  it('should move a completed task to pending list', () => {
    taskList[0].completed = true
    const wrapperMountC = mount(
      <ToDoAppList
        {...taskList}
        tasks={taskList}
        editTask={editTask}
        removeTask={removeTask}
      />,
    )
    wrapperMountC
      .find('#checkbox_1')
      .simulate('change', { target: { checked: false } })
    let pendingList = wrapperMountC.find('.taskTR_pending')
    expect(pendingList.length).toBe(2)
  })

  it('should move a pending task to completed list', () => {
    const wrapperMountC = mount(
      <ToDoAppList
        {...taskList}
        tasks={taskList}
        editTask={editTask}
        removeTask={removeTask}
      />,
    )
    wrapperMountC
      .find('#checkbox_1')
      .simulate('change', { target: { checked: true } })
    let completeList = wrapperMountC.find('.taskTR_completed')
    expect(completeList.length).toBe(1)

    const isCompleted = wrapperMountC
      .find('#checkbox_1')
      .hasClass('ToDo__checkbox_completed')
    expect(isCompleted).toBeTruthy()
  })

  it('should click delete Task button', () => {
    const wrapperMountRM = mount(
      <ToDoAppList
        tasks={taskList}
        editTask={editTask}
        removeTask={removeTask}
      />,
    )

    wrapperMountRM
      .find('AiFillDelete#removeBtn_2')
      .simulate('click', { id: '2' })
    let pendingList = wrapperMountRM.find('.taskTR_pending')
    expect(removeTask).toBeCalledTimes(1)
    expect(pendingList.length).toBe(1)
  })

  it('should call edit modal on click edit Button', () => {
    const wrapperEdit = mount(
      <ToDoAppList
        tasks={taskList}
        editTask={editTask}
        removeTask={removeTask}
      />,
    )
    wrapperEdit.find('FiEdit#editBtn_2').simulate('click', { id: '2' })
    expect(wrapperEdit.find(ToDoEditModal)).toHaveLength(1)
  })
})
