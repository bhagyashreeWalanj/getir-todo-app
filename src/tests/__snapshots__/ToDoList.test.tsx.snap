// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ToDoList Component should display to ToDo List Component 1`] = `
<div
  className="container"
>
  <div>
    <table
      className="table"
    >
      <thead>
        <tr>
          <th
            className="thead"
          >
            Status
          </th>
          <th
            className="thead"
          >
            Title
          </th>
          <th
            className="thead"
          >
            Created At
          </th>
          <th
            className="thead"
          >
            Task Status
          </th>
          <th
            className="thead"
          >
            Options
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            className="taskHeader"
          >
            Pending Tasks
          </td>
        </tr>
        <tr
          className="taskTR_pending"
          id="taskTR_pending_1"
          key="0"
        >
          <td>
            <input
              checked={false}
              className="ToDo__checkbox"
              id="checkbox_1"
              name="input_1"
              onChange={[Function]}
              type="checkbox"
              value="Hello World 1"
            />
          </td>
          <td
            className=""
          >
            Hello World 1
          </td>
          <td
            className=""
          >
            01/05/2022
          </td>
          <td
            className=""
          >
            PENDING
          </td>
          <td>
            <div
              className="options"
            >
              <FiEdit
                className="ToDo__update"
                id="editBtn_1"
                onClick={[Function]}
              >
                <span
                  className="tooltiptext"
                >
                  Edit
                </span>
              </FiEdit>
              <AiFillDelete
                className="ToDo__delete"
                id="removeBtn_1"
                onClick={[Function]}
              />
            </div>
          </td>
        </tr>
        <tr
          className="taskTR_pending"
          id="taskTR_pending_2"
          key="1"
        >
          <td>
            <input
              checked={false}
              className="ToDo__checkbox"
              id="checkbox_2"
              name="input_2"
              onChange={[Function]}
              type="checkbox"
              value="Hello World 2"
            />
          </td>
          <td
            className=""
          >
            Hello World 2
          </td>
          <td
            className=""
          >
            01/05/2022
          </td>
          <td
            className=""
          >
            PENDING
          </td>
          <td>
            <div
              className="options"
            >
              <FiEdit
                className="ToDo__update"
                id="editBtn_2"
                onClick={[Function]}
              >
                <span
                  className="tooltiptext"
                >
                  Edit
                </span>
              </FiEdit>
              <AiFillDelete
                className="ToDo__delete"
                id="removeBtn_2"
                onClick={[Function]}
              />
            </div>
          </td>
        </tr>
        <tr>
          <td
            className="taskHeader"
          >
            Completed Tasks
          </td>
        </tr>
        <tr
          className="noDataDiv"
        >
          <td>
            -
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
